Documentation for SHORTEST (#863)
neo4j/docs-cypher: modules/ROOT/pages/patterns/reference.adoc

A node pattern is a pattern that matches a single node.
It can be used on its own in a clause such as `MATCH` or `EXIST`, or form part of a xref:patterns/reference.adoc#path-patterns[path pattern]. 

See also xref:patterns/fixed-length-patterns.adoc#node-patterns[node pattern concepts].

[[node-patterns-syntax]]
=== Syntax

[source, syntax, role=noheader]
----
nodePattern ::= "(" [ nodeVariable ] [ labelExpression ] 
    [ propertyKeyValueExpression ] [ elementPatternWhereClause] ")"

> elementPatternWhereClause ::= "WHERE" booleanExpression

Reviewer: ```suggestion
    [ propertyKeyValueExpression ] [ elementPatternWhereClause] ")"

elementPatternWhereClause ::= "WHERE" booleanExpression
```

Added line break

----
relationshipPattern ::= fullPattern | abbreviatedRelationship

fullPattern ::=
    "<-[" patternFiller "]-"
  | "-[" patternFiller "]->"
  | "-[" patternFiller "]-"

abbreviatedRelationship ::= "<--" | "--" | "-->"

patternFiller ::= [ relationshipVariable ] [ typeExpression ] 
    [ propertyKeyValueExpression ] [ elementPatternWhereClause ]

> elementPatternWhereClause ::= "WHERE" booleanExpression

Reviewer: ```suggestion
    [ propertyKeyValueExpression ] [ elementPatternWhereClause ]

elementPatternWhereClause ::= "WHERE" booleanExpression
```

Added line break

[source, syntax, role=noheader]
----
pathPattern ::= 
  [ pathVariableDeclaration ] [ shortestPathSelector ] pathPatternExpression

pathVariableDeclaration ::= pathVariable "="
pathPatternExpression ::=
  { parenthesizedPathPatternExpression | pathPatternPhrase }

parenthesizedPathPatternExpression ::=
  "(" 
     [ subpathVariableDeclaration ]
     pathPatternExpression   
     [ parenthesizedPathPatternWhereClause ]
  ")"

subpathVariableDeclaration ::= pathVariable "="
pathPatternPhrase ::= [{ simplePathPattern | quantifiedPathPattern }]+
simplePathPattern ::= nodePattern 
  [ { relationshipPattern | quantifiedRelationship } nodePattern ]*

> parenthesizedPathPatternWhereClause ::= "WHERE" booleanExpression

Reviewer: ```suggestion
pathVariableDeclaration ::= pathVariable "="

pathPatternExpression ::=
  { parenthesizedPathPatternExpression | pathPatternPhrase }

parenthesizedPathPatternExpression ::=
  "(" 
     [ subpathVariableDeclaration ]
     pathPatternExpression   
     [ parenthesizedPathPatternWhereClause ]
  ")"

subpathVariableDeclaration ::= pathVariable "="

pathPatternPhrase ::= [{ simplePathPattern | quantifiedPathPattern }]+

simplePathPattern ::= nodePattern 
  [ { relationshipPattern | quantifiedRelationship } nodePattern ]*
  
parenthesizedPathPatternWhereClause ::= "WHERE" booleanExpression
```

Add line breaks.

[source, syntax, role=noheader]
----
quantifiedPathPattern ::= 
  parenthesizedPathPatternExpression quantifier

> fixedPath ::= nodePattern [ relationshipPattern nodePattern ]+

Reviewer: ```suggestion
  parenthesizedPathPatternExpression quantifier

fixedPath ::= nodePattern [ relationshipPattern nodePattern ]+
```

Added line break.

This syntax is still available.
It is equivalent to the syntax for quantified relationships, with the following differences:

* Position and syntax of quantifier.
* Semantics of the asterisk symbol.
* Type expressions are limited to the xref::syntax/operators.adoc#query-operators-boolean[disjunction operator].
* The xref:clauses/where.adoc[WHERE] clause is not allowed.

[[variable-length-relationships-syntax]]
=== Syntax

[source, syntax, role=noheader]
----
varLengthRelationship ::=
    "<-[" varLengthFiller "]-" 
  | "-[" varLengthFiller "]->"
  | "-[" varLengthFiller "]-"
varLengthFiller ::= [ relationshipVariable ] [ varLengthTypeExpression ] 
    [ varLengthQuantifier ] [ propertyKeyValueExpression ]

varLengthTypeExpression ::= ":" varLengthTypeTerm

varLengthTypeTerm ::=
    typeIdentifier
  | varLengthTypeTerm "|" varLengthTypeTerm

varLengthQuantifier ::= varLengthVariable | varLengthFixed

varLengthVariable ::= "*" [ [ lowerBound ] ".." [ upperBound ] ]

varLengthFixed ::= "*" fixedBound

fixedBound ::= unsignedInteger
lowerBound ::= unsignedInteger
upperBound ::= unsignedInteger

> unsignedInteger ::= [0-9]+

Reviewer: ```suggestion
varLengthRelationship ::=
    "<-[" varLengthFiller "]-" 
  | "-[" varLengthFiller "]->"
  | "-[" varLengthFiller "]-"

varLengthFiller ::= [ relationshipVariable ] [ varLengthTypeExpression ] 
    [ varLengthQuantifier ] [ propertyKeyValueExpression ]

varLengthTypeExpression ::= ":" varLengthTypeTerm

varLengthTypeTerm ::=
    typeIdentifier
  | varLengthTypeTerm "|" varLengthTypeTerm

varLengthQuantifier ::= varLengthVariable | varLengthFixed

varLengthVariable ::= "*" [ [ lowerBound ] ".." [ upperBound ] ]

varLengthFixed ::= "*" fixedBound

fixedBound ::= unsignedInteger

lowerBound ::= unsignedInteger

upperBound ::= unsignedInteger

unsignedInteger ::= [0-9]+
```

Added line breaks.

* `SHORTEST k`
* `ALL SHORTEST`
* `SHORTEST k GROUPS`
* `ANY k`
* `ALL`

The first three are shortest path selectors, and they specify which shortest paths should be returned, with the exact method of selection depending on the path selector type and the value of `k`.

See xref:patterns/reference.adoc#[path patterns] for details on where in the path pattern to include the path selector.

[[shortest-paths-syntax]]
=== Syntax

[source, syntax, role=noheader]
----
shortestPathSelector ::= 
  { allPathSearch | anyPathSearch | shortestPathSearch }

shortestPathSearch ::= 
  { allShortest | anyShortest | countedShortestPaths | 
     countedShortestGroups }

allShortest ::= "ALL SHORTEST" [ pathOrPaths ]

anyShortest ::= "ANY SHORTEST" [ pathOrPaths ]

countedShortestPaths ::= 
  "SHORTEST" numberOfPaths [ pathOrPaths ]

countedShortestGroups ::= 
  "SHORTEST" [ numberOfGroups ] [ pathOrPaths ] { "GROUP" | "GROUPS" }

allPathSearch ::= "ALL" [ pathOrPaths ]

anyPathSearch ::= "ANY" [ numberOfPaths ] [ pathOrPaths ]

pathOrPaths ::= { "PATH" | "PATHS" }

numberOfPaths ::= unsignedDecimalInteger

numberOfGroups ::= unsignedDecimalInteger

[ nodeVariable ] [ labelExpression ] 

>     [ propertyKeyValueExpression ] [ "WHERE" booleanExpression ] ")"

Reviewer: ```suggestion
```
I must have pasted these in error.

This is a _pre_-filter.
3. Paths are selected according to the path selector specified.
4. Any paths not satisfying the predicates in the graph pattern `WHERE` clause are removed.
This is a _post_-filter. 

.Examples of pre-filters
[source]
----
// node pattern WHERE clause
MATCH p = SHORTEST 2 (a WHERE a.p < 42)--+()

// relationship pattern property key-value expression
MATCH p = SHORTEST 2 (a)-[{p: 42}]-+()

// label and type expressions
MATCH p = SHORTEST 2 (:A)-[:R]-+(:B)

// quantified path pattern
MATCH SHORTEST 2 ((a)--(b) WHERE a.p < b.p)+

// parenthesized path pattern expression
// note the position of the parentheses!
MATCH SHORTEST 2 ( p = ()--+() WHERE any(n IN nodes(p) WHERE n.p < 42) )
----

.Examples of post-filters
[source]
----
// graph pattern WHERE clause
MATCH p = SHORTEST 2 (a)--+()
WHERE all(n IN nodes(p) WHERE n.p < 42)
----

[[shortest-paths-rules-partitions]]
==== Partitions

When a selective path selector is specified, then the paths matched by the path pattern are partitioned by distinct pairs of start and end nodes.
Within each partition, they are put in ascending order of path length.

> Within each partition, paths are then selected according to the specified path selector as follows:

Reviewer: ```suggestion
Within each partition, they are put in ascending order of path length, and then selected according to the specified path selector as follows:
```
I found the double 'Within each partition' jarring.

Reviewer: I see you have found a better solution.

[options="header",cols="a,3a"]
|===

| `SHORTEST k`
| First `k` paths, starting with the shortest.
Where more than one path could be picked for a given length, there is no particular order in which that selection is done.

> If `k` is greater than the number of paths in the partition, then all of the paths in that partition are returned.

Reviewer: This is being formatted as a header.

Author: I will add a new header: "Selective path selector" | "Partitioning"

Reviewer: Header works. But it has a typo. See separate comment.

Including either of them does not change path selection.
For example, the following two `MATCH` clauses are equivalent:

[source, role=noheader]
----
MATCH ALL SHORTEST PATHS (n:A) (()-->(:B))+
MATCH ALL SHORTEST (n:A) (()-->(:B))+
----

[[shortest-paths-examples]]
=== Examples

Return a single shortest path for each distinct pair of nodes matching `(:A)` and `(:B)`:

[source, role=noheader]
----
MATCH SHORTEST 1 (:A)-[:R]->{0,10}(:B)
----

Return any two paths connecting each distinct pair of nodes matching `(:A)` and `(:B)`:

[source, role=noheader]
----
MATCH p = ANY 2 (:A)-[:R]->{0,10}(:B)
----

Return all paths equal to the shortest path length for each distinct pair of nodes matching `(:A)` and `(:B)`:

[source, role=noheader]
----
MATCH ALL SHORTEST (:A)-[:R]->{0,10}(:B)
----

Return all paths equal to the two shortest path lengths for each distinct pair of nodes matching `(:A)` and `(:B)`:

[source, role=noheader]
----
MATCH SHORTEST 2 GROUPS (:A)-[:R]->{0,10}(:B)
----

Return a single shortest path for each distinct pair of nodes, where the path length is an even number:

[source, role=noheader]
----
MATCH SHORTEST 1 (p = ()--+() WHERE length(p) % 2 = 0)
----

> For every single shortest path connecting each distinct pair of nodes, only return those have path length that is an even number (so fewer results than the previous example):

Reviewer: ```suggestion
For every single shortest path connecting each distinct pair of nodes, only return those that have path length that is an even number (so fewer results than the previous example):
```

Both functions will continue to be available.

[[shortest-functions-syntax]]
=== Syntax

[source, syntax, role=noheader]
----
pathSelectorFunction ::= 
  { shortestPathFunction | allShortestPathsFunction }

shortestPathFunction ::= 
  "shortestPath(" + oneHopPathPatternExpression + ")"

allShortestPathsFunction ::= 
  "allShortestPaths(" + oneRelPathPatternExpression + ")"

oneRelPathPatternExpression ::= 
  nodePattern varLengthRelationship nodePattern
----

Note that it is possible to pass a fixed length path pattern (with a single relationship) to the path selector function, but doing so would not serve any purpose in discovering a shortest path.

[[shortest-functions-rules]]
=== Rules

[[shortest-functions-rules-var-length-restriction]]
==== Restricted to variable length

The pattern in the path selector function must be a variable-length relationship and not a quantified path pattern.

> These would not be valid:

Reviewer: ```suggestion
```

Reviewer: With the new code block headers, this seemed redundant.

.Not allowed
[source]
----
shortestPath(((a)-[:R]-(b)){1,5})
shortestPath((:A)-->+(:B))

> ----

Reviewer: ```suggestion

shortestPath((:A)-->+(:B))
----
```

[[shortest-functions-rules-path-pattern-length]]
==== Path pattern length

There must be exactly one relationship pattern in the path pattern.

> For example, this is valid:

Reviewer: ```suggestion
```

.Allowed
[source]
----
shortestPath((a)-[:R*1..5]-(b))
----

> But these are not:

Reviewer: ```suggestion
```

.Not allowed
[source]
----
shortestPath((a)-[:R*1..5]-(b)-->(:X))
shortestPath((:A))
allShortestPaths((a:A)-[:S*]->(:B), (a)-[:R*1..3]->(:C))

> ----

Reviewer: ```suggestion
shortestPath((a)-[:R*1..5]-(b)-->(:X))

shortestPath((:A))

allShortestPaths((a:A)-[:S*]->(:B), (a)-[:R*1..3]->(:C))
----
```

When the examples are lumped together, they look like they are part of the same query.

This is a _post_-filter.

.Examples of pre-filters
[source]
----
// node pattern WHERE clause
MATCH p = SHORTEST 2 (a WHERE a.p < 42)--+()

// relationship pattern property key-value expression
MATCH p = SHORTEST 2 (a)-[{p: 42}]-+()

// label and type expressions
MATCH p = SHORTEST 2 (:A)-[:R]-+(:B)

// quantified path pattern
MATCH SHORTEST 2 ((a)--(b) WHERE a.p < b.p)+

// parenthesized path pattern expression
// note the position of the parentheses!
MATCH SHORTEST 2 ( p = ()--+() WHERE any(n IN nodes(p) WHERE n.p < 42) )
----

.Examples of post-filters
[source]
----
// graph pattern WHERE clause
MATCH p = SHORTEST 2 (a)--+()
WHERE all(n IN nodes(p) WHERE n.p < 42)
----

[[shortest-paths-rules-partitions]]
==== Partitions

If there are multiple start and end nodes matching a path pattern, then, when a selective path selector is specified, the paths matched by the path pattern are partitioned by distinct pairs of start and end nodes.
Within each partition, they are put in ascending order of path length.
The partitioned paths are then selected according to the specified path selector as follows:

[options="header",cols="a,3a"]
|===

> | Selctive path selector | Partitioning

Reviewer: ```suggestion
| Selective path selector | Partitioning
```

[source, role=noheader]
----
MATCH SHORTEST 1 (:A)-[:R]->{0,10}(:B)
----

Return any two paths connecting each distinct pair of nodes matching `(:A)` and `(:B)`:

[source, role=noheader]
----
MATCH p = ANY 2 (:A)-[:R]->{0,10}(:B)
----

Return all paths equal to the shortest path length for each distinct pair of nodes matching `(:A)` and `(:B)`:

[source, role=noheader]
----
MATCH ALL SHORTEST (:A)-[:R]->{0,10}(:B)
----

Return all paths equal to the two shortest path lengths for each distinct pair of nodes matching `(:A)` and `(:B)`:

[source, role=noheader]
----
MATCH SHORTEST 2 GROUPS (:A)-[:R]->{0,10}(:B)
----

Return a single shortest path for each distinct pair of nodes, where the path length is an even number:

[source, role=noheader]
----
MATCH SHORTEST 1 (p = ()--+() WHERE length(p) % 2 = 0)
----

For every single shortest path connecting each distinct pair of nodes, only return those that have path length that is an even number (so fewer results than the previous example):

[source, role=noheader]
----
MATCH p = SHORTEST 2 ()--+()
WHERE length(p) % 2 = 0
----

[[shortest-functions]]
== The `shortestPath()` and `allShortestPaths()` functions

> Prior to the introduction of keyword-based specification of shortest path selection in Neo4j 5.17, the only available syntax for shortest paths were the functions `shortestPath()` and `allShortestPaths()`.

Reviewer 2: Let's not name the version before we have released the feature.
```suggestion
Prior to the introduction of keyword-based specification of shortest path selection in Neo4j 5.19, the only available syntax for shortest paths were the functions `shortestPath()` and `allShortestPaths()`.
```

----
shortestPath((a)-[:R*1..5]-(b))
----

.Not allowed
[source]
----
shortestPath((a)-[:R*1..5]-(b)-->(:X))

shortestPath((:A))

allShortestPaths((a:A)-[:S*]->(:B), (a)-[:R*1..3]->(:C))
----

[[shortest-functions-rules-filtering]]
==== Pre and post filtering

If the `MATCH` clause of the `shortestPath()` function includes a `WHERE` clause, this condition will act as a pre-filter: paths satisfying the `WHERE` clause are first found, and from those paths the shortest path is selected.

.Examples of pre-filters:
[source]
----
MATCH p = shortestPath(()-[*]-())
WHERE all(n in nodes(p) WHERE n.p < 42)

MATCH p = shortestPath(()-[*]-())
WHERE all(r in relationships(p) WHERE r.p < 42)
----

These contrast with `WHERE` clauses that are not part of the same `MATCH` clause.

.Example of post-filters
[source]
----
MATCH p = shortestPath(()-[*]-())
WITH nodes(p) AS N
WHERE all(n in N WHERE n.p < 42)
----

[[shortest-functions-rules-non-determinism]]
==== Non-determinism

If there is more than one path with minimum length, then `shortestPath()` function will return one of those paths non-deterministically.
As `allShortestPaths()` would return all of those paths, its results are deterministic.

> 

Reviewer 2: There is also a rule about `shortestPath()` not allowing to find shortest cycles, that is a shortest path from a node to itself.
